pkg/apis/resource/validation/validation.go

```diff
@@ -0,0 +1,325 @@
+				newItems := newSet.Difference(oldSet)
+				if len(newItems) > 0 {
+					allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "new entries may not get added while the claim is meant to be deallocated"))
+				}
+			}
+		}
+	}
+
+	if !oldClaim.Status.DeallocationRequested &&
+		resourceClaim.Status.DeallocationRequested &&
+		len(resourceClaim.Status.ReservedFor) > 0 {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("deallocationRequested"), "deallocation cannot be requested for claims which are in use"))
+	}
+
+	if resourceClaim.Status.Allocation == nil &&
+		resourceClaim.Status.DeallocationRequested {
+		if oldClaim.Status.Allocation != nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("deallocationRequested"), "must be reset together with removing the allocation"))
+		} else {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("deallocationRequested"), "claim is not allocated"))
+		}
+	}
+
+	// Once deallocation has been requested, that request cannot be removed
+	// anymore because the deallocation may already have started. The field
+	// can only get reset by the driver together with removing the
+	// allocation.
+	if oldClaim.Status.DeallocationRequested &&
+		!resourceClaim.Status.DeallocationRequested &&
+		resourceClaim.Status.Allocation != nil {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("deallocationRequested"), "deallocation cannot be canceled because it may already have started"))
+	}
+
+	return allErrs
+}
+
+func validateAllocationResult(allocation *resource.AllocationResult, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if allocation != nil {
+		if len(allocation.ResourceHandle) > resource.ResourceHandleMaxSize {
+			allErrs = append(allErrs, field.TooLongMaxLength(fldPath.Child("resourceHandle"), len(allocation.ResourceHandle), resource.ResourceHandleMaxSize))
+		}
+		if allocation.AvailableOnNodes != nil {
+			allErrs = append(allErrs, corevalidation.ValidateNodeSelector(allocation.AvailableOnNodes, fldPath.Child("availableOnNodes"))...)
+		}
+	}
+	return allErrs
+}
+
+func validateClaimUserReference(ref resource.ResourceClaimConsumerReference, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	// APIGroup is empty for the legacy group.
+	if ref.Resource == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("resource"), ""))
+	}
+	if ref.Name == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
+	}
+	if ref.UID == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("uid"), ""))
+	}
+	return allErrs
+}
+
+func validateResourceDriverName(driverName string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if driverName == "" {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	} else {
+		allErrs = append(allErrs, corevalidation.ValidateQualifiedName(strings.ToLower(driverName), fldPath)...)
```

Comment:
API comments said reverse-FQDN not a label-style name?

Reply:
The comment had it backwards.